sdks/java/extensions/arrow/src/main/java/org/apache/beam/sdk/extensions/arrow/ArrowConversion.java

```diff
@@ -0,0 +1,448 @@
+                    return FieldType.BYTES;
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.FixedSizeBinary type) {
+                    return FieldType.logicalType(FixedBytes.of(type.getByteWidth()));
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Bool type) {
+                    return FieldType.BOOLEAN;
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Decimal type) {
+                    // FieldType.DECIMAL isn't perfect here since arrow decimal has a
+                    // scale/precision fixed by the schema, but FieldType.DECIMAL uses a BigDecimal,
+                    // whose precision/scale can change from row to row.
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Date type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Time type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Timestamp type) {
+                    if (type.getUnit() == TimeUnit.MILLISECOND
+                        || type.getUnit() == TimeUnit.MICROSECOND) {
+                      return FieldType.DATETIME;
+                    } else {
+                      throw new IllegalArgumentException(
+                          "Unsupported timestamp unit: " + type.getUnit().name());
+                    }
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Interval type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+
+                  @Override
+                  public FieldType visit(ArrowType.Duration type) {
+                    throw new IllegalArgumentException(
+                        "Type \'" + type.toString() + "\' not supported.");
+                  }
+                });
+    return fieldType.withNullable(arrowFieldType.isNullable());
+  }
+
+  /**
+   * Returns an {@link Iterable<Row>} backed by the Arrow record batch stored in {@code
+   * vectorSchemaRoot}.
+   *
+   * <p>Note this is a lazy interface. The data in the underlying Arrow buffer is not read until a
+   * field of one of the returned {@link Row}s is accessed.
+   */
+  public static Iterable<Row> rowsFromRecordBatch(
+      Schema schema, VectorSchemaRoot vectorSchemaRoot) {
```

Comment:
you should clarify ownership of VectorSchemaRoot (i.e. will this iterator close after done?)

Reply:
+1 let's document the ownership model here. Let me know if you need help with this.